keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -0,0 +1,578 @@
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [x] (R) Graduation criteria is in place
+- [x] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+client-go's typed clients need a typesafe, programmatic way to make apply
+requests.
+
+## Motivation
+
+Currently, the only way to invoke server side apply from client-go is to call
+`Patch` with `PatchType.ApplyPatchType` and provide a `[]byte` containing the
+YAML or JSON of the apply configuration. This has a couple important
+deficiencies:
+
+- It is a gap completeness of the type client, which provides typesafe APIs for
+  all other major methods.
+- It makes it to easy for developers to make a major, but non-obvious mistake:
+  Use the existing go structs to construct an apply configuration, serialize
+  it to JSON, and pass it to `Patch`. This can cause zero valued required
+  fields being accientally included in the apply configuration resulting
+  in fields being accidentally set to incorrect values and/or fields accidentally
+  being accidentally being clamed as owned.
+
+Both sig-api-machinery and wg-api-expression agree that this enhancement is
+required for server side apply to be promoted to GA.
+
+### Goals
+
+Introduce a typesafe, programmatic way to call server side apply using the typed
+client in client-go.
+
+Express all apply configurations in Go that can be expressed in
+YAML. Specifically, an apply request must include only fields that are set by
+the applier and exclude those not set by the applier.
+
+Validate this enhancement meets the needs of developers:
+
+- An developer not directly involved in this enhancement successfully converts
+  a 1st party controller (one in github.com/kubernetes/kubernetes) to use this
+  enhancement.
+- A representative group of the developer community is made aware of this
+  proposed enhancement, is given early access to it via a fork of
+  controller-tools with the requisite generators, and is given the opportunity
+  to try it out and provide feedback.
+
+### Non-Goals
+
+Enhancements to client-go's dynamic client. The client-go dynamic client already
+supports Apply via Patch, which is adequate for the dynamic client.
```

Comment:
I agree that it's adequate for GA of the server-side-apply, but I suspect we'll want this fairly soon after we have it for other types.  Perhaps keep it as a non-goal of this kep, but indicate that some nicer mechanism is desired in the future?

Reply:
Sounds good. I've updated this section.